proposals/interface-types/Explainer.md

```diff
@@ -0,0 +1,663 @@
+another.
+
+With interface types available to use at the boundary between two
+modules, exports and imports can take, e.g., an abstract sequence of bytes,
+or a string, or a sequence of pairs of strings and integers, etc. With interface
+adapters to define how these abstract values are to be read from or written into
+linear memory, modules can use the wasm engine to take care of copying data
+between two modules' linear memories while allowing both modules to
+maintain full encapsulation.
+
+**Defining language-neutral interfaces like WASI**
+
+While many [WASI] signatures can be expressed in terms of `i32`s and,
+in the future, references to [type imports], there is still a need for WASI
+functions to take and return higher-level value types like strings or sequences.
+Currently, these values are specified in terms of linear memory. For example,
+in [`path_open`], a string is passed as a pair of `i32`s which represent the
+offset and length, respectively, of the string in linear memory.
+
+However, specifying this single, fixed representation for data types like
+strings will become problematic if:
+* WebAssembly gets access to GC memory with the [GC] proposal and the caller
+  wants to pass a `ref array u8`;
+* the host has a native string type which is not the same as the
+  GC `ref array u8` (e.g., a JS/DOM string) which WASI would ideally accept
+  without copy; or
+* the need arises to allow more than just UTF-8 encoding.
+
+With interface types, WASI functions can simply use the interface type `string`,
+allowing abstract string values to be created from linear memory, GC memory and
+host-string values, under the control of the calling module.
+
+
+## Overview
+
+This proposal builds on the following existing and separately proposed 
+high-level WebAssembly concepts:
+* **value type**: the set of types defined by the [core spec][Value Types]
+  that can be used to define globals, locals, functions, etc
+* **module**: the basic unit of WebAssembly code whose structure is defined in
+  the [core spec][Module Syntax] and which may only use value types in its
+  functions
+* **module type**: currently defined by the [core spec][Old Module Type] as a
+  mapping from imports to exports. More recently, there is a [proposal][New Module Type]
+  to generalize module types to include import/export names and have text
+  format parse rules so that module types can be written separately from
+  modules.
+* **known section**: sections defined by the [core spec's binary format][Known Section]
+  which collectively decode a module
+* **custom section**: sections defined by the [core spec's binary format][Custom Section]
+  whose contents are uninterpreted by the core spec but can be interpreted by
+  other tools or specifications (including this proposal)
+
+This proposal defines the following new high-level concepts:
+* **interface type**: a new set of types that describe abstract, high-level values
```

Comment:
Suggest rewording:
**interface type schema**: a system of types that allow the description of APIs in an abstract interoperable way without undue commitment to particular representations.

Reply:
Actually, maybe this is better:

**interface schema**: a system of types and values that allows  the description of APIs in an abstract way without undue commitment to particular representations.

Reply:
Even though the list is introduced by saying it's introducing "new high-level concepts", what the list's real purpose is is just defining terms that are used throughout the rest of the document (hence the boldface).  "Interface types" is used many times, and is symmetric to the existing term [Value types](https://webassembly.github.io/spec/core/syntax/types.html#value-types), so I think it has to stay: it's the term that refers to a specific set of types.  As you've defined "Interface schema", the term basically covers the whole proposal (types, values, instructions, validation rules, execution rules) and I don't see where in the explainer this specific term would currently be used.  But if you had a specific example, I'd be happy to discuss it.